bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1040,18 +1045,83 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
+        }
+
+        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+            return false;
+        }
+
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        BType targetRestType = targetType.getRestType();
+
+        BTupleType sourceTupleType;
+        if (sourceType.getTag() == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case UNSEALED:
+                    if (targetRestType == null) {
+                        return false;
+                    } else {
+                        return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                case CLOSED_SEALED:
+                    if (sourceArrayType.getSize() >= targetTypes.size()) {
+                        if (targetTypes.isEmpty()) {
+                            if (targetRestType != null) {
+                                return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                            }
+                            return true;
+                        }
+
+                        boolean isSameType = true;
+
+                        for (BType targetElementType : targetTypes) {
+                            if (!(checkIsType(sourceElementType, targetElementType, unresolvedTypes))) {
+                                isSameType = false;
+                            }
+                        }
+                        if (sourceArrayType.getSize() > targetTypes.size()) {
+                            if (targetRestType != null) {
+                                return (checkIsType (sourceElementType, targetRestType, unresolvedTypes) && isSameType);
+                            } else {
+                                return false;
+                            }
+                        }
+                        return isSameType;
+                    }
+                    break;
+            }
+
+            List<BType> fieldTypes = new ArrayList<>();
+            for (int i = 0; i < sourceArrayType.getSize(); i++) {
+                fieldTypes.add(sourceArrayType.getElementType());
+            }
+            sourceTupleType = new BTupleType(fieldTypes);
+        }
+
+        List<BType> sourceTypes = new ArrayList<>(sourceTupleType.getTupleTypes());
+        BType sourceRestType = sourceTupleType.getRestType();
+
+        if ((!(targetTypes.isEmpty()) && (sourceRestType != null && targetRestType == null))) {
             return false;
         }
 
-        List<BType> sourceTypes = new ArrayList<>(((BTupleType) sourceType).getTupleTypes());
-        BType sourceRestType = ((BTupleType) sourceType).getRestType();
         if (sourceRestType != null) {
```

Comment:
The logic from L1121 to L1131 doesn't seem right to me.

Where `x` is `[int, boolean, boolean...] x = [1, true]`,
`<any> x is [int, boolean...]` should evaluate to true.

But it doesn't because we seem to be doing a size check with the rest type also added as a separate type.